### PR TITLE
Aggs: no filter-by-filter if _doc_count field

### LINKED ISSUE #84048: Aggs: filter_by_filter optimization kicks in automatically to *slow down* `term` when `_doc_count` is present
### Elasticsearch Version

many

### Installed Plugins

none

### Java Version

_bundled_

### OS Version

all

### Problem Description

The `filter_by_filter` optimization should kick in when it makes aggs faster. We have a switch to turn it off, but we advise folks not to because it usually does a good job. Well, we've found a case where the heuristics do a bad job of picking the implementation - specifically I've seen a case when the `_doc_count` field exists and we're slower. I'm not really sure if this is universal, but it could be. I'm trying to figure out if this is an all the time thing or a sometimes thing.

### Steps to Reproduce

index docs with `_doc_count`. run a terms agg on a field with a cardinality in the ball park of 500.

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by nik9000**
The `_doc_count` field isn't great at random access and the
filter-by-filter execution mechanism for aggs takes one pass over the
data per bucket. This makes `_doc_count` very very slow in
filter-by-filter. So! This PR turns off the filter-by-filter execution
mode when there is more than one filter and there is a `_doc_count`
field.

Closes #84048

### CODE REVIEW

**Comment by nik9000** [comment]
@salvatore-campagna now that you are looking more into performance testing this might be worth looking at. I don't believe we have much where the `_doc_count` field is defined. And this really only comes up then.

**Comment by salvatore-campagna** [comment]
> @salvatore-campagna now that you are looking more into performance testing this might be worth looking at. I don't believe we have much where the `_doc_count` field is defined. And this really only comes up then.

@nik9000 What do you mean with "might be worth looking at". Are you suggesting to add a specific test for this use case?

**Comment by nik9000** [comment]
> Are you suggesting to add a specific test for this use case?

Yeah. Or, maybe, just more things with `_doc_count` set.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterByFilterAggregator.java [hunk 1]
```diff
@@ -109,6 +114,21 @@ final void add(QueryToFilterAdapter<?> filter) throws IOException {
...
             }
+            if (filters.size() == 1) {
+                /*
+                 * When we add the second filter we check if there are any _doc_count
+                 * fields and bail out of filter-by filter mode if there are. _doc_count
+                 * fields are expensive to decode and the overhead of iterating per
+                 * filter causes us to decode doc counts over and over again.
+                 */
+                Term term = new Term(DocCountFieldMapper.NAME, DocCountFieldMapper.NAME);
+                for (LeafReaderContext c : context.searcher().getLeafContexts()) {
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterByFilterAggregator.java:125]: I wonder if this is worth moving up to the `AggregationContext`.  I feel like a `hasDocCountField()` method might be useful beyond this, and I suspect it is intuitive to look for such a method on `AggregationContext`, but less so to look for it here.  Easy enough to do in a follow up if we want to though.

**Review by not-napoleon — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterByFilterAggregator.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterByFilterAggregator.java:125]: Makes sense to me.

**Pull request merged**